addon/adapters/osf-adapter.js

```diff
@@ -18,97 +20,256 @@ export default DS.JSONAPIAdapter.extend(HasManyQuery.RESTAdapterMixin, DataAdapt
         // Fix issue where CORS request failed on 301s: Ember does not seem to append trailing
         // slash to URLs for single documents, but DRF redirects to force a trailing slash
         var url = this._super(...arguments);
+        var options = (snapshot ? snapshot.adapterOptions : false) || {};
         if (requestType === 'deleteRecord' || requestType === 'updateRecord' || requestType === 'findRecord') {
             if (snapshot.record.get('links.self')) {
                 url = snapshot.record.get('links.self');
             }
+        } else if (options.url) {
+            url = options.url;
         }
+
         if (url.lastIndexOf('/') !== url.length - 1) {
             url += '/';
         }
         return url;
     },
     /**
-     * Construct a URL for a relationship create/update/delete. Has the same
-     * signature as buildURL, with the addition of a 'relationship' param
+     * Construct a URL for a relationship create/update/delete.
      *
      * @method _buildRelationshipURL
+     * @param {DS.Snapshot} snapshot
      * @param {String} relationship the relationship to build a url for
      * @return {String} a URL
      **/
-    _buildRelationshipURL(modelName, id, snapshot, requestType, query, relationship) { // jshint ignore:line
+    _buildRelationshipURL(snapshot, relationship) {
         var links = relationship ? snapshot.record.get(
             `relationshipLinks.${Ember.String.underscore(relationship)}.links`
         ) : false;
         if (links && (links.self || links.related)) {
             return links.self ? links.self.href : links.related.href;
+        }
+        return null;
+    },
+    /**
+     * Handle creation of related resources
+     *
+     * @param {DS.Store} store
+     * @param {DS.Snapshot} snapshot snapshot of inverse record
+     * @param {DS.Snapshot[]} createdSnapshots
+     * @param {String} relationship
+     * @param {String} url
+     * @param {Boolean} isBulk
+     **/
+    _createRelated(store, snapshot, createdSnapshots, relationship, url, isBulk = false) { // jshint ignore:line
+        // TODO support bulk create?
+        // if (isBulk) {
+        //
+        // }
+        return createdSnapshots.map(s => s.record.save({
```

Comment:
In certain circumstances, I receive an error:

> vendor.js:165498 Uncaught TypeError: createdSnapshots.map is not a function

Example: trying to add `z4esu` as a bibliographic read contributor from the node detail page (user management panel) after a refresh. If I try doing it again on the page, I get:
> vendor.js:30475 Uncaught Error: Assertion Failed: The id z4esu has already been used with another record of type dummy@model:contributor:.

The user is not added to the list (either visually, or after refreshing the page). It does not appear that a network request is sent.

This is the test project we've used before, so you should have access.